Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm

```diff
@@ -387,13 +405,123 @@ ResourceLoaderOptions loaderOptions(
+    if (!newPixelBuffer)
+        return false;
+
+    m_lastPixelBuffer = newPixelBuffer;
+
+    if (m_resourceOwner) {
+        if (auto surface = CVPixelBufferGetIOSurface(m_lastPixelBuffer.get()))
+            IOSurface::setOwnershipIdentity(surface, m_resourceOwner);
+    }
+
+    return true;
+}
+
+bool MediaPlayerPrivateWebM::updateLastImage()
+{
+    if (m_isGatheringVideoFrameMetadata) {
+        if (!m_lastPixelBuffer)
+            return false;
+        if (m_sampleCount == m_lastConvertedSampleCount)
+            return false;
+        m_lastConvertedSampleCount = m_sampleCount;
+    } else if (!updateLastPixelBuffer())
+        return false;
+
+    ASSERT(m_lastPixelBuffer);
+
+    if (!m_rgbConformer) {
+        auto attributes = @{ (__bridge NSString *)kCVPixelBufferPixelFormatTypeKey: @(kCVPixelFormatType_32BGRA) };
+        m_rgbConformer = makeUnique<PixelBufferConformerCV>((__bridge CFDictionaryRef)attributes);
+    }
+
+    m_lastImage = NativeImage::create(m_rgbConformer->createImageFromPixelBuffer(m_lastPixelBuffer.get()));
+    return true;
+}
+
+void MediaPlayerPrivateWebM::paint(GraphicsContext& context, const FloatRect& rect)
+{
+    paintCurrentFrameInContext(context, rect);
+}
+
+void MediaPlayerPrivateWebM::paintCurrentFrameInContext(GraphicsContext& context, const FloatRect& outputRect)
+{
+    if (context.paintingDisabled())
+        return;
+
+    auto image = nativeImageForCurrentTime();
+    if (!image)
+        return;
+
+    GraphicsContextStateSaver stateSaver(context);
+    FloatRect imageRect { FloatPoint::zero(), image->size() };
+    context.drawNativeImage(*image, imageRect.size(), outputRect, imageRect);
+}
+
+#if !HAVE(AVSAMPLEBUFFERDISPLAYLAYER_COPYDISPLAYEDPIXELBUFFER)
+void MediaPlayerPrivateWebM::willBeAskedToPaintGL()
+{
+    // We have been asked to paint into a WebGL canvas, so take that as a signal to create
+    // a decompression session, even if that means the native video can't also be displayed
+    // in page.
+    if (m_hasBeenAskedToPaintGL)
+        return;
+
+    ALWAYS_LOG(LOGIDENTIFIER);
+    m_hasBeenAskedToPaintGL = true;
+    acceleratedRenderingStateChanged();
+}
+#endif
+
+RefPtr<VideoFrame> MediaPlayerPrivateWebM::videoFrameForCurrentTime()
+{
+    if (!m_isGatheringVideoFrameMetadata)
+        updateLastPixelBuffer();
+    if (!m_lastPixelBuffer)
+        return nullptr;
+    return VideoFrameCV::create(currentMediaTime(), false, VideoFrame::Rotation::None, RetainPtr { m_lastPixelBuffer });
+}
+
+DestinationColorSpace MediaPlayerPrivateWebM::colorSpace()
+{
+    updateLastImage();
+    return m_lastImage ? m_lastImage->colorSpace() : DestinationColorSpace::SRGB();
+}
+
 void MediaPlayerPrivateWebM::setNaturalSize(FloatSize size)
 {
     FloatSize oldSize = m_naturalSize;
     m_naturalSize = size;
     if (oldSize != m_naturalSize) {
         INFO_LOG(LOGIDENTIFIER, "was ", oldSize.width(), " x ", oldSize.height(), ", is ", size.width(), " x ", size.height());
```

Comment:
`FloatSize` has a `LogArgument` template, so you should be able to write this as 

`        INFO_LOG(LOGIDENTIFIER, "was ", oldSize, ", is ", size);`